pyscf/dft/dft_parser.py

```diff
@@ -0,0 +1,76 @@
+
+# Copyright 2014-2024 The PySCF Developers. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+'''
+unified dft parser for coordinating dft protocols with
+1. xc functionals
+2. dispersion corrections / nonlocal correction
+3. GTO basis (TODO)
+4. geometrical counterpoise (gCP) correction (TODO)
+'''
+
+from functools import lru_cache
+
+# supported dispersion corrections
+DISP_VERSIONS = ['d3bj', 'd3zero', 'd3bjm', 'd3zerom', 'd3op', 'd4']
+
+@lru_cache(128)
+def parse_dft(dft_method):
+    ''' conventional dft method ->
+    (xc, enable nlc, dispersion, with 3-body dispersion)
+    '''
+    if not isinstance(dft_method, str):
+        return dft_method, None, None, False
+    method_lower = dft_method.lower()
+    xc = dft_method
+    disp = None
+
+    # special cases:
+    # - wb97x-d is not supported yet
+    # - wb97*-d3bj is wb97*-v with d3bj
+    # - wb97x-d3 is not supported yet
+    # - 3c method is not supported yet
+
+    if method_lower == 'wb97x-d':
+        raise NotImplementedError('wb97x-d is not supported yet.')
+
+    if method_lower == 'wb97m-d3bj':
+        return 'wb97m-v', False, ('wb97m', 'd3bj', False)
+    if method_lower == 'b97m-d3bj':
+        return 'b97m-v', False, ('b97m', 'd3bj', False)
+    if method_lower == 'wb97x-d3bj':
+        return 'wb97x-v', False, ('wb97x', 'd3bj', False)
+
+    # J. Chem. Theory Comput. 2013, 9, 1, 263–272
+    if method_lower in ['wb97x-d3']:
+        raise NotImplementedError('wb97x-d3 is not supported yet.')
+
+    if method_lower.endswith('-3c'):
+        raise NotImplementedError('*-3c methods are not supported yet.')
+
+    for d in DISP_VERSIONS:
```

Comment:
May need to consider a corner case like `wb97m-d3bjm`

Reply:
`wb97m-d3bjm` is not a well-defined xc functional. In this parser, `wb97m-d3bjm` will be interpreted as `wb97m` and `d3bjm`. `wb97m` is not a well-defined xc functional. I am assuming xc parser or libxc will raise an error.